catroid/src/main/java/org/catrobat/catroid/transfers/project/ProjectUploadService.kt

```diff
@@ -0,0 +1,206 @@
+    }
+
+    private fun createUploadNotification(programName: String): Notification {
+        StatusBarNotificationManager.getInstance().createNotificationChannel(applicationContext)
+
+        var uploadIntent = Intent(applicationContext, MainMenuActivity::class.java)
+        uploadIntent.action = Intent.ACTION_MAIN
+        uploadIntent = uploadIntent.setFlags(Intent.FLAG_ACTIVITY_RESET_TASK_IF_NEEDED)
+        val pendingIntent = PendingIntent.getActivity(
+            applicationContext,
+            StatusBarNotificationManager.UPLOAD_PENDING_INTENT_REQUEST_CODE,
+            uploadIntent, PendingIntent.FLAG_CANCEL_CURRENT
+        )
+
+        return NotificationCompat.Builder(applicationContext, StatusBarNotificationManager.CHANNEL_ID)
+            .setContentIntent(pendingIntent)
+            .setSmallIcon(R.drawable.ic_stat)
+            .setContentTitle(
+                applicationContext.getString(R.string.notification_upload_title_pending) + " " + programName
+            )
+            .setContentText(applicationContext.getString(R.string.notification_upload_pending))
+            .setOngoing(true)
+            .setProgress(MAX_PERCENT, 0, true)
+            .build()
+    }
+
+    private fun createUploadFinishedNotification(programName: String): Notification {
+        var uploadIntent = Intent(applicationContext, MainMenuActivity::class.java)
+        uploadIntent.action = Intent.ACTION_MAIN
+        uploadIntent = uploadIntent.setFlags(Intent.FLAG_ACTIVITY_RESET_TASK_IF_NEEDED)
+        val pendingIntent = PendingIntent.getActivity(
+            applicationContext,
+            StatusBarNotificationManager.UPLOAD_PENDING_INTENT_REQUEST_CODE,
+            uploadIntent, PendingIntent.FLAG_CANCEL_CURRENT
+        )
+
+        return NotificationCompat.Builder(applicationContext, StatusBarNotificationManager.CHANNEL_ID)
+            .setContentIntent(pendingIntent)
+            .setSmallIcon(R.drawable.ic_stat)
+            .setContentTitle(
+                applicationContext.getString(R.string.notification_upload_title_finished) + " " + programName
+            )
+            .setContentText(applicationContext.getString(R.string.notification_upload_finished))
+            .setAutoCancel(true)
+            .setOngoing(false)
+            .setProgress(0, 0, false)
+            .build()
+    }
+
+    private fun getUserEmail(provider: String): String {
+        val sharedPreferences = PreferenceManager.getDefaultSharedPreferences(applicationContext)
+
+        val email = when (provider) {
+            FACEBOOK -> sharedPreferences.getString(FACEBOOK_EMAIL, NO_FACEBOOK_EMAIL)
+            GOOGLE_PLUS -> sharedPreferences.getString(GOOGLE_EMAIL, NO_GOOGLE_EMAIL)
+            else -> sharedPreferences.getString(EMAIL, NO_EMAIL)
+        }
+
+        val result = if (email.equals(NO_EMAIL)) {
+            DeviceSettingsProvider.getUserEmail(this)
+        } else {
+            email
+        }
+
+        return result ?: ""
```

Comment:
There is no check email == null for value result. Am I right, that this is an implicit second check of the email? Why is NO_EMAIL handled different than a null email here?

Reply:
Well, the only way we can get a null email in the result value here is that somebody put a null value for any of the keys into the shared preferences or we cannot find a google account on the device in case no email has been provided by the user. 

The `?:` operator is called elvis operator btw (see https://kotlinlang.org/docs/reference/null-safety.html#elvis-operator) and is used here to return `result` or, if it is `null`, an empty string.